src/PoolManager.sol

```diff
@@ -203,34 +214,62 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
-
-        uint256 feeForProtocol;
-        uint24 swapFee;
-        Pool.SwapState memory state;
-        (delta, feeForProtocol, swapFee, state) = pools[id].swap(
-            Pool.SwapParams({
-                tickSpacing: key.tickSpacing,
-                zeroForOne: params.zeroForOne,
-                amountSpecified: params.amountSpecified,
-                sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
-        );
-
-        _accountPoolBalanceDelta(key, delta);
-
-        // the fee is on the input currency
-        unchecked {
-            if (feeForProtocol > 0) {
-                protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+        int128 hookDeltaInSpecified = key.hooks.beforeSwap(key, params, hookData);
+
+        bool exactInput = params.amountSpecified < 0;
+        bool zeroIsSpecified = exactInput == params.zeroForOne;
+
+        // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+        params.amountSpecified += hookDeltaInSpecified;
+        if (exactInput ? params.amountSpecified > 0 : params.amountSpecified < 0) revert SwapTypeChanged();
+
+        // Account the hook's delta to the hook's address
+        _accountDeltaFor(zeroIsSpecified ? key.currency0 : key.currency1, hookDeltaInSpecified, address(key.hooks));
+
+        {
+            uint256 feeForProtocol;
+            uint24 swapFee;
+            Pool.SwapState memory state;
+            (delta, feeForProtocol, swapFee, state) = pools[id].swap(
+                Pool.SwapParams({
+                    tickSpacing: key.tickSpacing,
+                    zeroForOne: params.zeroForOne,
+                    amountSpecified: params.amountSpecified,
+                    sqrtPriceLimitX96: params.sqrtPriceLimitX96
+                })
+            );
+
+            // the fee is on the input currency
+            unchecked {
+                if (feeForProtocol > 0) {
+                    protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+                }
             }
+
+            // TODO returning deltas here doesnt make any sense
```

Comment:
it really feels like this should be returning the final deltas as of line 266

Reply:
I saw @danrobinson's PR on event order, but if the swap takes place over the course of the before/after hooks, it is technically completed after any recursive swap calls by the hook, so I actually think it's ok to make this event at the end of the function after the after hook. It's recursive so it makes sense, inner calls complete before you work your way back up to the outer calls...

Reply:
Right one is the update of the pool's concentrated liquidity and one is the update of the actual user's experience... i think that realistically this is a bigger question that won't be addressed in this PR but across all events at a later date. Cody and the routing team are all going to have input here, and talking to integrators and indexers etc to see what they need our events to include.

Reply:
Ugh yeah this is tough for sure. Let's raise this tomorrow?

cc: @codyborn

Reply:
As one input, at least for how I use event data in Dune, I would be pretty confused if the sqrtPriceX96 and tick of the most recent event fired didn't match the current state of the pool